lib/Promise.php

```diff
@@ -101,9 +101,9 @@ public function __construct(callable $executor = null) {
      * @param callable $onRejected
      * @return Promise
      */
-    public function then(callable $onFulfilled = null, callable $onRejected = null) {
+    function then(callable $onFulfilled = null, callable $onRejected = null) {
 
-        $subPromise = new Promise();
+        $subPromise = new self();
```

Comment:
:-1: new static is in general a bad idea. For the same reason I stated earlier the last time we had a discussion about it. It makes assumptions about the behavior and the constructor the subclass will have.